Unstructured Mesh - Initial implementation

pshriwise opened the pull request:
## Summary
This PR leverages the mesh library underlying DAGMC to provide unstructured mesh tally support, which allows us to provide spatial tallies that conform to specific regions of the problem geometry. Specifically, this implementation supports collision and tracklength tallies on tetrahedral meshes (see obligatory images for a feature of this nature below).

### Pin fluxes in a bare PWR assembly model
![image](https://user-images.githubusercontent.com/4563941/77113811-73df2b00-69f9-11ea-9bff-6e6fbe56b08c.png)
### Flux in a piping manifold modeled using a combination of DAGMC geometry and an unstructured mesh tally.
![image](https://user-images.githubusercontent.com/4563941/77113866-89eceb80-69f9-11ea-9cc2-e5699f2a3425.png)
## Use
An unstructured mesh filter can be applied in the same way that our structured mesh options work currently in either pure XML or via the Python API. The main difference being that it stores a filename to be loaded at runtime rather than the information necessary to describe the mesh itself.
### XML
```
<?xml version='1.0' encoding='utf-8'?>
<tallies>
  <mesh id="1" type="unstructured">
    <mesh_file>manifold.h5m</mesh_file>
  </mesh>
  <filter id="1" type="mesh">
    <bins>1</bins>
  </filter>
  <filter id="2" type="energy">
    <bins>0.0 1.0 10000000.0</bins>
  </filter>
  <tally id="1">
    <filters>1 2</filters>
    <scores>flux</scores>
    <estimator>tracklength</estimator>
  </tally>
</tallies>
```
### Python
```python
umesh = openmc.UnstructuredMesh(filename="pins1-4.h5m")
mesh_filter = openmc.MeshFilter(umesh)
tally = openmc.Tally()
tally.filters = [mesh_filter]
tally.scores = ['heating', 'flux']
tally.estimator = 'tracklength'
model.tallies = (tally,)
```
## Notable Changes
  - Introduction of a C++ `UnstructuredMesh` class
  - Introduction of a Python `UnstructuredMesh` class
  - Addition of a `score_int_to_str` function
  - A jupyter notebook applying an unstructured mesh to a reactor problem
  - A jupyter notebook applying an unstructured mesh to a fixed source problem using CAD geometry
  - Documentation updates related to this capability
## Visualization
For tallies using _only_ an unstructured mesh filter, a VTK (viewable in VisIt or Paraview) file is produced with every statepoint file, resulting in flies with the naming scheme `tally_<tally_id>.<batch_number>.vtk`. These files contain two datasets for each tally score/nuclide combination -- one with the normalized tally value and one with the standard deviation of the tally value.
I opted not to write this file if other filters are present on the tally due to the rapidly increasing number of possible datasets that would then need to be written to the VTK file for completeness.
Instead, the volumes and centroids of the unstructured mesh elements are provided in the statepoint file and can be used to generate an alternative visualization of some subset of the tally data after it has been extracted using the Python API.
## Testing
A series of 8 tests have been added which compare the results of an unstructured tetrahedral mesh where groups of the tetrahedra align with cells of a regular mesh that is also present in the run. The results of these tetrahedral groups are summed and compared to the result of the corresponding regular mesh cell for verification. This is done with a geometry going beyond the bounds of the mesh as well as one aligned with the boundary and tetrahedral meshes with missing groups of tets to ensure the tally scoring routing handles tracks exiting and re-entering the mesh appropriately.

## Performance

Using this feature incurs a pretty big hit in performance. There are a couple of factors that contribute to this:

  1. It's simply more expensive to perform the necessary operations in an unstructured mesh. This implementation employs a KD-tree to improve this search, but it is still quite slow compared to our structured mesh classes.

  2. Some scalability issues with MOAB have been observed, with a cap on increasing particle rates at around 16-20 threads for the machines I've tested this on. I'm in discussions with their team on how to address this currently.

Overall, I'm hoping to improve these issues in the future and that we're ok with the trade-off for the added capability for now.

## Future work
I was tempted to label this PR "Unstructured Mesh - Part 1 of ?" as there are a number of subsequent PRs that will follow this one related to unstructured mesh support including:

  - Improved visualization options in the Python API
  - Support for collision estimator tallies for arbitrary mesh element types with [libmesh](https://libmesh.github.io/)
  - Addition of a translation attribute to the `MeshFilter` class to allow unstructured mesh templates to be used in multiple locations with little additional cost.

makeclean commented:
@shimwell will be happy about this one :)
________________________________
From: Patrick Shriwise <notifications@github.com>
Sent: Friday, March 20, 2020 4:06:16 PM
To: openmc-dev/openmc <openmc@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: [openmc-dev/openmc] Unstructured Mesh - Initial implementation (#1527)

Summary

This PR leverages the mesh library underlying DAGMC to provide unstructured mesh tally support, which allows us to provide spatial tallies that conform to specific regions of the problem geometry. Specifically, this implementation supports collision and tracklength tallies on tetrahedral meshes (see obligatory images for a feature of this nature below).

Pin fluxes in a bare PWR assembly model

[image]<https://user-images.githubusercontent.com/4563941/77113811-73df2b00-69f9-11ea-9bff-6e6fbe56b08c.png>

Flux in a piping manifold modeled using a combination of DAGMC geometry and an unstructured mesh tally.

[image]<https://user-images.githubusercontent.com/4563941/77113866-89eceb80-69f9-11ea-9cc2-e5699f2a3425.png>

Use

An unstructured mesh filter can be applied in the same way that our structured mesh options work currently in either pure XML or via the Python API. The main difference being that it stores a filename to be loaded at runtime rather than the information necessary to describe the mesh itself.

XML

<?xml version='1.0' encoding='utf-8'?>
<tallies>
  <mesh id="1" type="unstructured">
    <mesh_file>manifold.h5m</mesh_file>
  </mesh>
  <filter id="1" type="mesh">
    <bins>1</bins>
  </filter>
  <filter id="2" type="energy">
    <bins>0.0 1.0 10000000.0</bins>
  </filter>
  <tally id="1">
    <filters>1 2</filters>
    <scores>flux</scores>
    <estimator>tracklength</estimator>
  </tally>
</tallies>


Python

umesh = openmc.UnstructuredMesh(filename="pins1-4.h5m")
mesh_filter = openmc.MeshFilter(umesh)
tally = openmc.Tally()
tally.filters = [mesh_filter]
tally.scores = ['heating', 'flux']
tally.estimator = 'tracklength'
model.tallies = (tally,)

Notable Changes

  *   Introduction of a C++ UnstructuredMesh class
  *   Introduction of a Python UnstructuredMesh class
  *   Addition of a score_int_to_str function
  *   A jupyter notebook applying an unstructured mesh to a reactor problem
  *   A jupyter notebook applying an unstructured mesh to a fixed source problem using CAD geometry
  *   Documentation updates related to this capability

Visualization

For tallies using only an unstructured mesh filter, a VTK (viewable in VisIt or Paraview) file is produced with every statepoint file, resulting in flies with the naming scheme tally_<tally_id>.<batch_number>.vtk. These files contain two datasets for each tally score/nuclide combination -- one with the normalized tally value and one with the standard deviation of the tally value.
I opted not to write this file if other filters are present on the tally due to the rapidly increasing number of possible datasets that would then need to be written to the VTK file for completeness.
Instead, the volumes and centroids of the unstructured mesh elements are provided in the statepoint file and can be used to generate an alternative visualization of some subset of the tally data after it has been extracted using the Python API.

Testing

A series of 8 tests have been added which compare the results of an unstructured tetrahedral mesh where groups of the tetrahedra align with cells of a regular mesh that is also present in the run. The results of these tetrahedral groups are summed and compared to the result of the corresponding regular mesh cell for verification. This is done with a geometry going beyond the bounds of the mesh as well as one aligned with the boundary and tetrahedral meshes with missing groups of tets to ensure the tally scoring routing handles tracks exiting and re-entering the mesh appropriately.

Performance

Using this feature incurs a pretty big hit in performance. There are a couple of factors that contribute to this:

  1.  It's simply more expensive to perform the necessary operations in an unstructured mesh. This implementation employs a KD-tree to improve this search, but it is still quite slow compared to our structured mesh classes.

  2.  Some scalability issues with MOAB have been observed, with a cap on increasing particle rates at around 16-20 threads for the machines I've tested this on. I'm in discussions with their team on how to address this currently.

Overall, I'm hoping to improve these issues in the future and that we're ok with the trade-off for the added capability for now.

Future work

I was tempted to label this PR "Unstructured Mesh - Part 1 of ?" as there are a number of subsequent PRs that will follow this one related to unstructured mesh support including:

  *   Improved visualization options in the Python API
  *   Support for collision estimator tallies for arbitrary mesh element types with libmesh<https://libmesh.github.io/>
  *   Addition of a translation attribute to the MeshFilter class to allow unstructured mesh templates to be used in multiple locations with little additional cost.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/openmc-dev/openmc/pull/1527

Commit Summary

  *   Starting definition of general mesh interface.
  *   Finishing inheritance infrastructure.
  *   Adding method for finding bins_crossed.
  *   Building tree manually.
  *   Finishing concrete implementation.
  *   Filling in call to kdtree for computing track lengths.
  *   Updating check for unstructured mesh.
  *   Taking care of a couple setup issues.
  *   Adding to mesh abstraction to support other operations external to the mesh source files.
  *   Wrapping more umesh stuff into ifdefs. Moving pointers into vector during initialization.
  *   Updates to get reg mesh working as an abstracted class
  *   Removing unstructured mesh vector.
  *   Using a common count_sites implementation.
  *   Making common function for laying ray over unstructured mesh.
  *   adding compute barycentric data call.
  *   First successfull umesh tally, maybe.
  *   Some edits, fixes, documentation.
  *   Removing silly cast.
  *   Adding body to the Unstructured mesh to_hdf5 method.
  *   Removing volume normalization from tally values.
  *   Writing total unstructured mesh volume to statepoint as well.
  *   Corrections after rebase.
  *   Adding ifdefs for mesh object IO.
  *   Some cleanup and removal of commented sections after rebase.
  *   Adding unstructured mesh class.
  *   Can read unstructured mesh from statepoint now.
  *   Some (temporary) changes to make unstructured mesh copacetic with the other mesh bin structures.
  *   Correction to the check location for partial intersections with tets at the end of a track.
  *   USE DAGMC EXPORTED TARGETS
  *   Initializing internal volume parameter and using len() to size bins (so we aren't assuming use of a numpy array)
  *   Factoring out a new SructuredMesh class. Plot method placement still needs to be resolved.
  *   A couple fixes.
  *   Writing out element centroids to the statepoint file.
  *   Adding umesh tally write between batches when accumulating.
  *   Writing scores between each batch.
  *   Setting all unstructured mesh tally data/error values at once during accumulate.
  *   Some additions/improvements after the rebase.
  *   Another fix to this pesky data writing problem - need contiguous memory.
  *   Initial test for unstructured mesh.
  *   Including tests w external geometry and holes in the mesh.
  *   Removing .dat files
  *   Renaming regular mesh variables for clarity.
  *   Updating score tagging for more meaningful values on the mesh.
  *   Moving tally writing into the statepoint function. Should write anytime a statepoint file is asked for.
  *   Writing unstructured mesh results with statepoint.
  *   Supporting output of multiple scores on unstructured mesh.
  *   Adding warning for tallies with more than one filter. Printing output upon writing the unstructured mesh file.
  *   Adding by score and nuclide now.
  *   Using score and nuclide strings for a nicer tag name.
  *   Added a function for converting score integers to strings instead of storing them.
  *   Including break to skip mesh file write.
  *   Start of code cleanup.
  *   Doc string for attribute and a format update.
  *   Adding centroids attribute.
  *   Writing to VTK by default as it is more universal.
  *   Adding the unstructured mesh example notebook.
  *   Adding unstructured mesh with CAD example.
  *   Updates to the unstructured mesh documentation.
  *   Additions to documentation for file specs and examples.
  *   Correct spelling error in output.
  *   Some updates to the current documentation for accuracy.
  *   Updating unstructured mesh description.
  *   Fixing bracket.
  *   Some updates to the example file.
  *   Updating example notebooks to get rid of a weird library error.
  *   Some cleanup of the unstructured mesh source.
  *   Some added documentation/cleanup of Python API files.
  *   Some cleanup of the writing function during statepoint output and removing unecessary changes in the eigenvalue file.
  *   Various cleanup changes.
  *   Matching xtl xtensor versions with develop.
  *   Adding comments for clarity.

File Changes

  *   M docs/source/examples/index.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-076b4ac5ec5e90a2de8fe695e1138fe0> (12)
  *   A docs/source/examples/unstructured-mesh-part-i.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-f77ccb5fc1d80f96c667bf70a3da950a> (13)
  *   A docs/source/examples/unstructured-mesh-part-ii.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-41d5c8f25609f3c7b25705ca44668a10> (13)
  *   M docs/source/io_formats/statepoint.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-42ecc33bf7caf2d033e973e37f553527> (4)
  *   M docs/source/io_formats/tallies.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-ec3d3889e8dd9113cfe40a0f93f6e84c> (33)
  *   M docs/source/pythonapi/base.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a0587741a66d5d898838029d6aa95ab4> (1)
  *   M docs/source/usersguide/beginners.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-89adefcf098dbb3cd4a0074a0e8d74f5> (2)
  *   M docs/source/usersguide/processing.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-3150f1d616242bc55bdc39d8fabc975a> (6)
  *   M docs/source/usersguide/settings.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-2615d1fa896e7277e8a9629f0387fb3f> (17)
  *   A examples/jupyter/images/manifold-cad.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-80de89627580b9ae8aa254d79a8bf6fe> (0)
  *   A examples/jupyter/images/manifold_flux.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-0207e8bcfda81a6504a0311e31e8cebc> (0)
  *   A examples/jupyter/images/manifold_pnt_cld.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-99cbc7cfaaf76494c2176470d982d179> (0)
  *   A examples/jupyter/images/pin_mesh.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-b3f0ba41f730e5c7ba60b71aa4091212> (0)
  *   A examples/jupyter/images/umesh_flux.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a3aedb6f9e08219df7a07759e0f2e747> (0)
  *   A examples/jupyter/images/umesh_heating.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-254fe9f081b4cd5b03a1857564cea12d> (0)
  *   A examples/jupyter/images/umesh_w_assembly.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a67c8d5685add9c382b22b267e84fb3c> (0)
  *   A examples/jupyter/unstructured-mesh-part-i.ipynb<https://github.com/openmc-dev/openmc/pull/1527/files#diff-3051a19f68f7920f18b1b6a3ba3a8e7d> (946)
  *   A examples/jupyter/unstructured-mesh-part-ii.ipynb<https://github.com/openmc-dev/openmc/pull/1527/files#diff-c5e3989528344cf85619f88d8ed51889> (1331)
  *   M include/openmc/mesh.h<https://github.com/openmc-dev/openmc/pull/1527/files#diff-48008356a9bfd6ebd5d1d96bb9eb726c> (256)
  *   M include/openmc/state_point.h<https://github.com/openmc-dev/openmc/pull/1527/files#diff-16c998f29b38bbc665fd2d140f450f85> (4)
  *   M include/openmc/tallies/tally.h<https://github.com/openmc-dev/openmc/pull/1527/files#diff-420a02c1e379f3b720d6eb500cc630e7> (3)
  *   M openmc/filter.py<https://github.com/openmc-dev/openmc/pull/1527/files#diff-50afb764b870f7f00586a2f65cf16ae7> (5)
  *   M openmc/mesh.py<https://github.com/openmc-dev/openmc/pull/1527/files#diff-dff0eee6a53f775079d50598371be4eb> (109)
  *   M src/mesh.cpp<https://github.com/openmc-dev/openmc/pull/1527/files#diff-9a5625e73a45d7ece1335647e0fa27d6> (651)
  *   M src/state_point.cpp<https://github.com/openmc-dev/openmc/pull/1527/files#diff-19a6222f68e840db270338ba445e6d51> (79)
  *   M src/tallies/filter_mesh.cpp<https://github.com/openmc-dev/openmc/pull/1527/files#diff-9ae0dac0a2609e434897d80aeb68d55b> (13)
  *   M src/tallies/tally.cpp<https://github.com/openmc-dev/openmc/pull/1527/files#diff-323e0b60215e7a8fe92fe52359a79d38> (174)
  *   A tests/regression_tests/unstructured_mesh/__init__.py<https://github.com/openmc-dev/openmc/pull/1527/files#diff-d2f95e48e513b2a6be0ae02866768fd8> (0)
  *   A tests/regression_tests/unstructured_mesh/test.py<https://github.com/openmc-dev/openmc/pull/1527/files#diff-d48e6c6a1811a908c69afc22dd1c8f70> (270)
  *   A tests/regression_tests/unstructured_mesh/test_mesh_tets.h5m<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a9bbeca9352cf5c90084447da3bf6f08> (0)
  *   A tests/regression_tests/unstructured_mesh/test_mesh_tets_w_holes.h5m<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a9d1ca0bb1f68c10c2529b5d13caba38> (0)

Patch Links:

  *   https://github.com/openmc-dev/openmc/pull/1527.patch
  *   https://github.com/openmc-dev/openmc/pull/1527.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/openmc-dev/openmc/pull/1527>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AASTUSVFTBHLPTFIUU6UOBLRIOH7RANCNFSM4LQPISWA>.

Shimwell commented:
> @Shimwell will be happy about this one :)
> […](#)
> ________________________________ From: Patrick Shriwise <notifications@github.com> Sent: Friday, March 20, 2020 4:06:16 PM To: openmc-dev/openmc <openmc@noreply.github.com> Cc: Subscribed <subscribed@noreply.github.com> Subject: [openmc-dev/openmc] Unstructured Mesh - Initial implementation (#1527) Summary This PR leverages the mesh library underlying DAGMC to provide unstructured mesh tally support, which allows us to provide spatial tallies that conform to specific regions of the problem geometry. Specifically, this implementation supports collision and tracklength tallies on tetrahedral meshes (see obligatory images for a feature of this nature below). Pin fluxes in a bare PWR assembly model [image]<https://user-images.githubusercontent.com/4563941/77113811-73df2b00-69f9-11ea-9bff-6e6fbe56b08c.png> Flux in a piping manifold modeled using a combination of DAGMC geometry and an unstructured mesh tally. [image]<https://user-images.githubusercontent.com/4563941/77113866-89eceb80-69f9-11ea-9cc2-e5699f2a3425.png> Use An unstructured mesh filter can be applied in the same way that our structured mesh options work currently in either pure XML or via the Python API. The main difference being that it stores a filename to be loaded at runtime rather than the information necessary to describe the mesh itself. XML <?xml version='1.0' encoding='utf-8'?> <tallies> <mesh id="1" type="unstructured"> <mesh_file>manifold.h5m</mesh_file> </mesh> <filter id="1" type="mesh"> <bins>1</bins> </filter> <filter id="2" type="energy"> <bins>0.0 1.0 10000000.0</bins> </filter> <tally id="1"> <filters>1 2</filters> <scores>flux</scores> <estimator>tracklength</estimator> </tally> </tallies> Python umesh = openmc.UnstructuredMesh(filename="pins1-4.h5m") mesh_filter = openmc.MeshFilter(umesh) tally = openmc.Tally() tally.filters = [mesh_filter] tally.scores = ['heating', 'flux'] tally.estimator = 'tracklength' model.tallies = (tally,) Notable Changes * Introduction of a C++ UnstructuredMesh class * Introduction of a Python UnstructuredMesh class * Addition of a score_int_to_str function * A jupyter notebook applying an unstructured mesh to a reactor problem * A jupyter notebook applying an unstructured mesh to a fixed source problem using CAD geometry * Documentation updates related to this capability Visualization For tallies using only an unstructured mesh filter, a VTK (viewable in VisIt or Paraview) file is produced with every statepoint file, resulting in flies with the naming scheme tally_<tally_id>.<batch_number>.vtk. These files contain two datasets for each tally score/nuclide combination -- one with the normalized tally value and one with the standard deviation of the tally value. I opted not to write this file if other filters are present on the tally due to the rapidly increasing number of possible datasets that would then need to be written to the VTK file for completeness. Instead, the volumes and centroids of the unstructured mesh elements are provided in the statepoint file and can be used to generate an alternative visualization of some subset of the tally data after it has been extracted using the Python API. Testing A series of 8 tests have been added which compare the results of an unstructured tetrahedral mesh where groups of the tetrahedra align with cells of a regular mesh that is also present in the run. The results of these tetrahedral groups are summed and compared to the result of the corresponding regular mesh cell for verification. This is done with a geometry going beyond the bounds of the mesh as well as one aligned with the boundary and tetrahedral meshes with missing groups of tets to ensure the tally scoring routing handles tracks exiting and re-entering the mesh appropriately. Performance Using this feature incurs a pretty big hit in performance. There are a couple of factors that contribute to this: 1. It's simply more expensive to perform the necessary operations in an unstructured mesh. This implementation employs a KD-tree to improve this search, but it is still quite slow compared to our structured mesh classes. 2. Some scalability issues with MOAB have been observed, with a cap on increasing particle rates at around 16-20 threads for the machines I've tested this on. I'm in discussions with their team on how to address this currently. Overall, I'm hoping to improve these issues in the future and that we're ok with the trade-off for the added capability for now. Future work I was tempted to label this PR "Unstructured Mesh - Part 1 of ?" as there are a number of subsequent PRs that will follow this one related to unstructured mesh support including: * Improved visualization options in the Python API * Support for collision estimator tallies for arbitrary mesh element types with libmesh<https://libmesh.github.io/> * Addition of a translation attribute to the MeshFilter class to allow unstructured mesh templates to be used in multiple locations with little additional cost.
> ________________________________ You can view, comment on, or merge this pull request online at: #1527 Commit Summary * Starting definition of general mesh interface. * Finishing inheritance infrastructure. * Adding method for finding bins_crossed. * Building tree manually. * Finishing concrete implementation. * Filling in call to kdtree for computing track lengths. * Updating check for unstructured mesh. * Taking care of a couple setup issues. * Adding to mesh abstraction to support other operations external to the mesh source files. * Wrapping more umesh stuff into ifdefs. Moving pointers into vector during initialization. * Updates to get reg mesh working as an abstracted class * Removing unstructured mesh vector. * Using a common count_sites implementation. * Making common function for laying ray over unstructured mesh. * adding compute barycentric data call. * First successfull umesh tally, maybe. * Some edits, fixes, documentation. * Removing silly cast. * Adding body to the Unstructured mesh to_hdf5 method. * Removing volume normalization from tally values. * Writing total unstructured mesh volume to statepoint as well. * Corrections after rebase. * Adding ifdefs for mesh object IO. * Some cleanup and removal of commented sections after rebase. * Adding unstructured mesh class. * Can read unstructured mesh from statepoint now. * Some (temporary) changes to make unstructured mesh copacetic with the other mesh bin structures. * Correction to the check location for partial intersections with tets at the end of a track. * USE DAGMC EXPORTED TARGETS * Initializing internal volume parameter and using len() to size bins (so we aren't assuming use of a numpy array) * Factoring out a new SructuredMesh class. Plot method placement still needs to be resolved. * A couple fixes. * Writing out element centroids to the statepoint file. * Adding umesh tally write between batches when accumulating. * Writing scores between each batch. * Setting all unstructured mesh tally data/error values at once during accumulate. * Some additions/improvements after the rebase. * Another fix to this pesky data writing problem - need contiguous memory. * Initial test for unstructured mesh. * Including tests w external geometry and holes in the mesh. * Removing .dat files * Renaming regular mesh variables for clarity. * Updating score tagging for more meaningful values on the mesh. * Moving tally writing into the statepoint function. Should write anytime a statepoint file is asked for. * Writing unstructured mesh results with statepoint. * Supporting output of multiple scores on unstructured mesh. * Adding warning for tallies with more than one filter. Printing output upon writing the unstructured mesh file. * Adding by score and nuclide now. * Using score and nuclide strings for a nicer tag name. * Added a function for converting score integers to strings instead of storing them. * Including break to skip mesh file write. * Start of code cleanup. * Doc string for attribute and a format update. * Adding centroids attribute. * Writing to VTK by default as it is more universal. * Adding the unstructured mesh example notebook. * Adding unstructured mesh with CAD example. * Updates to the unstructured mesh documentation. * Additions to documentation for file specs and examples. * Correct spelling error in output. * Some updates to the current documentation for accuracy. * Updating unstructured mesh description. * Fixing bracket. * Some updates to the example file. * Updating example notebooks to get rid of a weird library error. * Some cleanup of the unstructured mesh source. * Some added documentation/cleanup of Python API files. * Some cleanup of the writing function during statepoint output and removing unecessary changes in the eigenvalue file. * Various cleanup changes. * Matching xtl xtensor versions with develop. * Adding comments for clarity. File Changes * M docs/source/examples/index.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-076b4ac5ec5e90a2de8fe695e1138fe0> (12) * A docs/source/examples/unstructured-mesh-part-i.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-f77ccb5fc1d80f96c667bf70a3da950a> (13) * A docs/source/examples/unstructured-mesh-part-ii.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-41d5c8f25609f3c7b25705ca44668a10> (13) * M docs/source/io_formats/statepoint.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-42ecc33bf7caf2d033e973e37f553527> (4) * M docs/source/io_formats/tallies.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-ec3d3889e8dd9113cfe40a0f93f6e84c> (33) * M docs/source/pythonapi/base.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a0587741a66d5d898838029d6aa95ab4> (1) * M docs/source/usersguide/beginners.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-89adefcf098dbb3cd4a0074a0e8d74f5> (2) * M docs/source/usersguide/processing.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-3150f1d616242bc55bdc39d8fabc975a> (6) * M docs/source/usersguide/settings.rst<https://github.com/openmc-dev/openmc/pull/1527/files#diff-2615d1fa896e7277e8a9629f0387fb3f> (17) * A examples/jupyter/images/manifold-cad.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-80de89627580b9ae8aa254d79a8bf6fe> (0) * A examples/jupyter/images/manifold_flux.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-0207e8bcfda81a6504a0311e31e8cebc> (0) * A examples/jupyter/images/manifold_pnt_cld.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-99cbc7cfaaf76494c2176470d982d179> (0) * A examples/jupyter/images/pin_mesh.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-b3f0ba41f730e5c7ba60b71aa4091212> (0) * A examples/jupyter/images/umesh_flux.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a3aedb6f9e08219df7a07759e0f2e747> (0) * A examples/jupyter/images/umesh_heating.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-254fe9f081b4cd5b03a1857564cea12d> (0) * A examples/jupyter/images/umesh_w_assembly.png<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a67c8d5685add9c382b22b267e84fb3c> (0) * A examples/jupyter/unstructured-mesh-part-i.ipynb<https://github.com/openmc-dev/openmc/pull/1527/files#diff-3051a19f68f7920f18b1b6a3ba3a8e7d> (946) * A examples/jupyter/unstructured-mesh-part-ii.ipynb<https://github.com/openmc-dev/openmc/pull/1527/files#diff-c5e3989528344cf85619f88d8ed51889> (1331) * M include/openmc/mesh.h<https://github.com/openmc-dev/openmc/pull/1527/files#diff-48008356a9bfd6ebd5d1d96bb9eb726c> (256) * M include/openmc/state_point.h<https://github.com/openmc-dev/openmc/pull/1527/files#diff-16c998f29b38bbc665fd2d140f450f85> (4) * M include/openmc/tallies/tally.h<https://github.com/openmc-dev/openmc/pull/1527/files#diff-420a02c1e379f3b720d6eb500cc630e7> (3) * M openmc/filter.py<https://github.com/openmc-dev/openmc/pull/1527/files#diff-50afb764b870f7f00586a2f65cf16ae7> (5) * M openmc/mesh.py<https://github.com/openmc-dev/openmc/pull/1527/files#diff-dff0eee6a53f775079d50598371be4eb> (109) * M src/mesh.cpp<https://github.com/openmc-dev/openmc/pull/1527/files#diff-9a5625e73a45d7ece1335647e0fa27d6> (651) * M src/state_point.cpp<https://github.com/openmc-dev/openmc/pull/1527/files#diff-19a6222f68e840db270338ba445e6d51> (79) * M src/tallies/filter_mesh.cpp<https://github.com/openmc-dev/openmc/pull/1527/files#diff-9ae0dac0a2609e434897d80aeb68d55b> (13) * M src/tallies/tally.cpp<https://github.com/openmc-dev/openmc/pull/1527/files#diff-323e0b60215e7a8fe92fe52359a79d38> (174) * A tests/regression_tests/unstructured_mesh/__init__.py<https://github.com/openmc-dev/openmc/pull/1527/files#diff-d2f95e48e513b2a6be0ae02866768fd8> (0) * A tests/regression_tests/unstructured_mesh/test.py<https://github.com/openmc-dev/openmc/pull/1527/files#diff-d48e6c6a1811a908c69afc22dd1c8f70> (270) * A tests/regression_tests/unstructured_mesh/test_mesh_tets.h5m<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a9bbeca9352cf5c90084447da3bf6f08> (0) * A tests/regression_tests/unstructured_mesh/test_mesh_tets_w_holes.h5m<https://github.com/openmc-dev/openmc/pull/1527/files#diff-a9d1ca0bb1f68c10c2529b5d13caba38> (0) Patch Links: * https://github.com/openmc-dev/openmc/pull/1527.patch * https://github.com/openmc-dev/openmc/pull/1527.diff — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub<#1527>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AASTUSVFTBHLPTFIUU6UOBLRIOH7RANCNFSM4LQPISWA>.

Nomination for best new feature award.

Shimwell commented:
Nomination for best new feature.

lekeOnilude commented:
jjncncnnnc

pshriwise commented on docs/source/usersguide/beginners.rst:
```diff
@@ -60,7 +60,7 @@ Now let's look at the pros and cons of Monte Carlo methods:
 - **Pro**: No mesh generation is required to build geometry. By using
   `constructive solid geometry`_, it's possible to build arbitrarily complex
-  models with curved surfaces.
+  reactor models with curved surfaces.
```
I was trying to clarify that for reactor geometries CSG is almost always sufficient, but that other applications might need something more (e.g. CAD).

Rather than specify "reactor" here, which does seem kinda random reading it now, I'd probably remove "arbitrarily" from this sentence instead.

pshriwise commented on examples/jupyter/unstructured-mesh-part-i.ipynb:
```diff
@@ -0,0 +1,946 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAvAAAARgCAYAAACPEGIDAAAABHNCSVQICAgIfAhkiAAAABl0RVh0U29mdHdhcmUAZ25vbWUtc2NyZWVuc2hvdO8Dvz4AACAASURBVHic7J15l11VtfbXdvjfO8Zr71WxVxAEVFqB0IS+FUJa/ISmh2AHBEKCBEEQRBEQveq1u+LF7v0C9/2jzjw1z3OeOdfapyqp2ns/vzEyzt5rtvucVPKsWauS7v/8n//zv0UIIYQQQggxCN631Q0IIYQQQggh2pGAF0IIIYQQYkBIwAshhBBCCDEg3t913Vb3IIQQQgghhGhEE3ghhBBCCCEGhAS8EEIIIYQQA0ICXgghhBBCiAGhM/BCCCGEEEI ...
```
Ah, yeah for sure. Meant to clear those files from the other notebook out before committing this - must've missed it in the last round.

pshriwise commented on examples/jupyter/unstructured-mesh-part-i.ipynb:
```diff
@@ -0,0 +1,946 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAvAAAARgCAYAAACPEGIDAAAABHNCSVQICAgIfAhkiAAAABl0RVh0U29mdHdhcmUAZ25vbWUtc2NyZWVuc2hvdO8Dvz4AACAASURBVHic7J15l11VtfbXdvjfO8Zr71WxVxAEVFqB0IS+FUJa/ISmh2AHBEKCBEEQRBEQveq1u+LF7v0C9/2jzjw1z3OeOdfapyqp2ns/vzEyzt5rtvucVPKsWauS7v/8n//zv0UIIYQQQggxCN631Q0IIYQQQggh2pGAF0IIIYQQYkBIwAshhBBCCDEg3t913Vb3IIQQQgghhGhEE3ghhBBCCCEGhAS8EEIIIYQQA0ICXgghhBBCiAGhM/BCCCGEEEI ...
```
Haha fair point! I'll pull that.

pshriwise commented on include/openmc/mesh.h:
```diff
@@ -218,13 +242,189 @@ class RectilinearMesh : public Mesh
+  //! \param[in] r1 Ending position
+  //! \param[out] ijk Indices of the mesh bin containing the intersection point
+  //! \return Whether the line segment connecting r0 and r1 intersects mesh
+  bool intersects(Position& r0, Position r1, int* ijk);
```
:grimacing: Thanks for catching that.

pshriwise commented on include/openmc/mesh.h:
```diff
@@ -218,13 +242,189 @@ class RectilinearMesh : public Mesh
+private:
+  moab::Range ehs_; //!< Range of tetrahedra EntityHandle's in the mesh
+  moab::EntityHandle meshset_; //!< EntitySet containing all elements
+  moab::EntityHandle tet_set_; //! < EntitySet containing all tetrahedra
```
Hrmmm good point. I thought at one point that having all of the elements (triangles included) around, might be useful I guess. Moving to `tetset_` only.

pshriwise commented on src/mesh.cpp:
```diff
@@ -1502,6 +1534,597 @@ openmc_mesh_set_params(int32_t index, int n, const double* ll, const double* ur,
+      lengths.push_back((hit.first - last_dist) / track_len);
+    } else {
+      // if in the loop, we should always find a tet
+      warning("No tet found for location between triangle hits");
```
In a glancing intersection where a track intersects with the vertex of a tet (or directly on an edge of a face ) where we might not find a containing tet b/c it lies right on the boundary of that element. The presumption here is that this is ok because the track length in that element would be near-zero, but I thought if there were something systematically wrong it would be best to let the user know if this is happening frequently.

As the comment above this line indicates, I originally thought it should be an error.

pshriwise commented on src/mesh.cpp:
```diff
@@ -1502,6 +1534,597 @@ openmc_mesh_set_params(int32_t index, int n, const double* ll, const double* ur,
+    }
+
+    moab::CartVect p[4];
+    rval = mbi_->get_coords(&(verts[0]), (int)verts.size(), p[0].array());
```
`data()` Makes these calls much cleaner. Thanks!

pshriwise replied:
And the c-style cast isn't necessary. My bad.

pshriwise commented on src/mesh.cpp:
```diff
@@ -1502,6 +1534,597 @@ openmc_mesh_set_params(int32_t index, int n, const double* ll, const double* ur,
+  }
+
+  // compute the centroid of the element vertices
+  moab::CartVect centroid(0.0);
```
Initializes all values of the `CartVect` to zero. The empty constructor doesn't guarantee that :confounded:

pshriwise commented on src/mesh.cpp:
```diff
@@ -1502,6 +1534,597 @@ openmc_mesh_set_params(int32_t index, int n, const double* ll, const double* ur,
+  rval = mbi_->write_mesh(filename.c_str(), &tet_set_, 1);
+  if (rval != moab::MB_SUCCESS) {
+    std::stringstream msg;
+    msg << "Failed to write unstructured mesh " << id_;
```
Feeling like I went through and made these changes at some point and lost them :disappointed:

pshriwise commented on src/state_point.cpp:
```diff
@@ -677,6 +683,79 @@ void read_source_bank(hid_t group_id)
+  for (auto& tally : model::tallies) {
+    for (auto filter_idx : tally->filters()) {
+      auto& filter = model::tally_filters[filter_idx];
+      if (filter->type() == "mesh") {
```
Nice. I like it!

pshriwise commented on src/state_point.cpp:
```diff
@@ -677,6 +683,79 @@ void read_source_bank(hid_t group_id)
+              for (int j = 0; j < tally->results_.shape()[0]; j++) {
+                double mean = tally->results_(j, nuc_score_idx, TallyResult::SUM) / n_realizations;
+                double sum_sq = tally->results_(j , nuc_score_idx, TallyResult::SUM_SQ);
+                double std_dev = sum_sq / n_realizations - std::pow(mean, 2) / (n_realizations - 1);
```
Thanks for catching these. I was writing out the sum-of-squares originally and switched to std. dev. right at the end. Should've checked it more carefully. And you're right, I made this change to provide more meaningful values for visualization purposes. 

These values are all going into the VTK file though. The sums/sums-of-squares are still written to the statepoint file and are available during a restart.

pshriwise commented on src/tallies/tally.cpp:
```diff
@@ -65,6 +65,171 @@ double global_tally_collision;
 double global_tally_leakage;
 
+std::string
+score_int_to_str(int score_int) {
```
B/c I'm a fool and didn't see that function. Much easier. Thanks!

pshriwise commented:
Re: `UnstructuredMesh` class naming

A subsequent PR that adds support for libmesh does just this. I didn't end up making that change until adding support for libmesh later on. If you're ok with it, I'll plan on including it then.

pshriwise commented on tests/regression_tests/unstructured_mesh/test.py:
```diff
@@ -0,0 +1,270 @@
+        settings.export_to_xml()
+
+    def _compare_inputs(self):
+        pass
```
I added this because the xml's are going to differ depending on the test being run as a result of using `pytest`'s parameterize here. The unstructured mesh filename and tally estimator (maybe other values too?) in the `tally.xml` will be updated depending on the set of options passed to the `UnstructuredMeshTest` class. 

There's probably a better solution to this though -- bypassing a standard test function like isn't great it's true.

pshriwise replied:
In my last commit here, I've submitted a possible solution for this which allows the `PyAPITestHarness` to take in a filename for the input check which it will use to both update and check the test results.

pshriwise commented:
Given that there were some significant updates in the last few commits, I feel I should explain the changes a bit.

I was hoping to use a method for tracklength tallies that uses the intersected triangles to walk from tet to tet in the mesh, tallying the length between each intersection appropriately. The motivation for that being it would avoid searched for containing tets (i.e. do another search in the KDTree data structure) as we tally each segment of the track in the mesh. Unfortunately, the corner cases we can encounter are too problematic. Coincident intersections with 2-3 triangles for tracks passing through a vertex and tracks that pass through the edges of tets create cases that require a lot of deduction and branching to resolve with a lot of potential for mistakes and further corner cases. I thought I had these worked out in a fairly clean fashion. Cleanup of the test file revealed otherwise (I wasn't testing what I thought I was - thanks to @paulromano's suggestions for revealing that) and our accuracy in comparison to values for the structured mesh comparison was pretty poor with relative differences over 100% for some bins in the test cases.

In the end, I went back to an implementation used previously by @makeclean in DAGMC's tally library I had originally implemented when starting this endeavor. This method finds the midpoint of each intersection segment, does a search for a containing tet using the midpoint, and tallies the normalized segment length to the bin for that tet if one is found. It's simple, clean, accurate, and simplifies some of the data structures that were needed for the "mesh walk" method. The only tweak I made to the "midpoint" method was to remove coincident intersection distances from the set of intersections returned from the KDTree's `ray_intersect_triangles` call. They don't do any harm, but we end up doing a lot of extra work to add zeros to the set of tally scores returned from the `bins_crossed` method. This method improved the accuracy of the tracklength tally greatly (comparison to the `RegularMesh` matches within 8 decimal places) with a modest hit in performance (anecdotally it looks like ~15-20%).

paulromano commented on include/openmc/mesh.h:
```diff
@@ -218,13 +241,178 @@ class RectilinearMesh : public Mesh
+                    std::vector<int>& bins,
+                    std::vector<double>& lengths) const override;
+
+private:
```
[NL.16](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rl-order): Use `public` before `protected` before `private` (there's currently a second public section below this)

paulromano commented on openmc/mesh.py:
```diff
@@ -586,3 +588,129 @@ def to_xml_element(self):
+    name : str
+        Name of the mesh
+    filename : str
+        Location of the unstructured mesh file
```
```suggestion
    filename : str
        Location of the unstructured mesh file
    mesh_id : int
        Unique identifier for the mesh
    name : str
        Name of the mesh
```

paulromano commented on openmc/mesh.py:
```diff
@@ -586,3 +588,129 @@ def to_xml_element(self):
+        Unique identifier for the mesh
+    name : str
+        Name of the mesh
+    mesh_file : str
```
```suggestion
    filename : str
```
Also, it does seem slightly odd to me that in the HDF5 file we are using `mesh_file` whereas here we use `filename`.

pshriwise replied:
True. I'll update that to be more consistent.

paulromano commented on src/mesh.cpp:
```diff
@@ -1502,6 +1525,550 @@ openmc_mesh_set_params(int32_t index, int n, const double* ll, const double* ur,
+UnstructuredMesh::bins_crossed(const Particle* p,
+                               std::vector<int>& bins,
+                               std::vector<double>& lengths) const {
+
```
```suggestion
                               std::vector<double>& lengths) const
{
```

pshriwise commented:
Thanks @paulromano! 

For some reason, GitHub had all of the suggested changes marked as outdated and wouldn't allow me to apply them as usual. I've added them manually and addressed other comments in the latest commit. Let me know if I missed anything!

paulromano commented:
Great new feature, thanks @pshriwise!
